EIPS/eip-draft_request_id.md

```diff
@@ -0,0 +1,155 @@
+
+*Citing from the [`BlockHeaders` spec definition](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#blockheaders-0x04):* 
+
+>`[blockHeader_0, blockHeader_1, ...]`
+
+>Reply to `GetBlockHeaders`. The items in the list (following the message ID) are block
+headers in the format described in the main Ethereum specification, previously asked for
+in a GetBlockHeaders message. This may validly contain no block headers if none of the
+requested block headers were found. The number of headers that can be requested in a
+single message may be subject to implementation-defined limits.
+
+Let's consider a client making many simultanous requests for `GetBlockHeaders` to one of its peers. By nature it can not be guaranteed that the expected responses arrive in the same order as they were sent. For the client to associate the incoming responses to the correct requests it has to loop through all pending requests trying to match it with the incoming response based on its contents.
+
+This can be particular tricky for responses that are ambigious such as empty responses.
+
+Now consider we change both the `GetBlockHeaders` and the `BlockHeaders` command to include a `request_id`.
+
+* `GetBlockHeaders (0x03)`
+   * **Current:** `[block: {P, B_32}, maxHeaders: P, skip: P, reverse: P in {0, 1}]`
+   * **Then**: `[request_id: P, [block: {P, B_32}, maxHeaders: P, skip: P, reverse: P in {0, 1}]]`
+* `BlockHeaders (0x04)`
+   * **Current:** `[blockHeader_0, blockHeader_1, ...]`
+   * **Then**: `[request_id: P, [blockHeader_0, blockHeader_1, ...]]`
+
+The `request_id` is a simple integer set by the client when it makes the request. On the responding side, the exact same `request_id` from the incoming request is put back into the response object.
+
+This change allows the requesting client to match incoming responses **directly** back to their pending requests without going through all of the pending requests to check if they might match based on the response data.
+
+The selected request/response pair serves as an example for many similar request/response pairs in the `eth` networking protocol.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+The lack of request identifiers in the request / response paris of the `eth` protocol puts unnecessary burden of code complexity into every Ethereum client. It also makes the communication slightly less efficient. Another argument can be made that the addition of request identifiers makes the protocol more aligned with the `les` protocol which **does** already defines request identifiers for each request / response pair.
```

Comment:
This might be the place to mention that:
- the extra complexity only occurs if you want to implement pipelining.
- because of this complexity neither geth or parity have implemented pipelining
- pipelining allows nodes to sync off each other faster, which is a likely result of this EIP being accepted.

Reply:
Interesting! I'm not exactly sure if I understand the benefits beyond "We can make more parallel requests to node `A` for any given command e.g. `GetBlockHeaders`". That is mentioned in the EIP but I think you are talking about different benefits.

Btw you should feel free to add commits to the PR and add yourself as a co-author if you like :)